### ISSUE TITLE
Split long compiles across muiltiple updates

### ISSUE THREAD

**Issue by Dunbaratu** [KSP-KOS/KOS#1053]
Currently when executing a `run program.`, if `program` is anything other than a very quick little script, you can noticeably see the universe freeze for a bit while it finishes the compile.  This happens because the entire compilation must fit inside one pass of FixedUpdate(), and the system thinks of it as only taking one single instruction out of the IPU's.  Until the compile finishes, FixedUpdate() doesn't return, and the entire KSP game is frozen.
### Implementation idea:

To fix this, compilation should probably use the Unity Coroutine system.  When the compile starts, we begin a co-routine to execute the compile inside of, and the compiler checks itself with a stopwatch to yield after it's been too long, choosing to yield until the next FixedUpdate() to pick up and continue where it left off.

In the meantime, the main FixedUpdate() for that CPU that's doing the compile should do nothing until the compile coroutine is done.  It should know to bypass its work until the compile says its done.  Otherwise the program would end up continuing on past the `run` command while the `run` is still working, and that would break everything.

**Comment by Dunbaratu** [comment]
Okay I think that co-routines may be entirely messed up too much here to use.  They are implemented on top of C#'s IEnumerable, and use the IEnumerable's ``yield return`` keyword as a sort of weird indirect way to create the threading (you have to make the coroutine return an IEnumerable, and Unity uses the thing you ``yield return`` to decide when to call you again.  If you ``yield return WaitForFixedUpdate();`` for example, then the system won't call you again until next fixedupdate to get the "next iteration" value from your "iterator".  If your real goal is just to suspend execution at a certain point in an *existing* system with deeply nested method calls inside method calls inside method calls, like our compiler, it gets really messy as you now have to go through and rewrite all the method calls in that chain to instead return IEnumerable, so they can all ``yield return`` in a nested chain all the way back up to where you started the compiler from.  Since these are method calls that already return useful OTHER things, you end up having to move their real return values to ``out`` parameters instead because the coroutine system demands that they must all return IEnumerable, to have them pass the ``yield return`` up the chain the outermost method that began the compile.

At least that's how I think it works.  I can find no good examples of how the heck you use `yield return` in a method nesting way to interrupt your activity in a method call nested inside a method call nested inside the one you actually called.

As far as I can tell, it looks like the ``yield return`` statement must occur in the outermost nested method, which makes the whole system useless when the goal was to pause execution somewhere deeply nested inside the compiler, like by putting a Stopwatch check inside VisitNode() to have it notice when it's time to yield.

That's going to be too ugly, I think.

### alternate solution - just use plain old normal threading?

Given that the compiler lives entirely inside kOS.Safe, it may be possible to just execute it in a thread without having that thread break Unity or KSP because kOS.Safe never touches Unity code.

If that simpler solution can work, then all we really need to do is set an 'is compiling" flag, kick off the compile in a thread that sets "is compiling" to false when it's done, and then have FixedUpdate() return and do nothing as long as "is compiling" is still on.

**Comment by Dunbaratu** [comment]
Given that a while back @hvacengi took over the work for threading, on which this is based, I changed the assignee on the issue to him.

**Comment by marianoapp** [comment]
Actually the delay happens in the parsing phase due to Mono's poor implementation of case insensitive regexs (at least with the version shipped with KSP, but that could change now that they are moving to Unity 5).
During my tests the parsing took 10+ times when case insensitivity was enabled, and that was what prompted me to implement the lowercasing algorithm (which was reverted on #925 due to valid complaints).

**Comment by Dunbaratu** [comment]
Oh, interesting. I never knew that.  Your code worked well, but wasn't documented for the follow-on programmer who took the work over, so I often had to guess what your intentions were.  In fact I never knew you added the lowercasing yourself - I thought it was a leftover from Nivekk.  Nor did I know that it was a workaround to avoid regex'ing it in the parser generator.  It just looked like a weird poor design choice instead of a deliberate compromise.  Usually when I know I'm doing something oddball that looks "wrong" to most programmers, that's when I make sure to comment it well, to protect against someone coming along, seeing it, and trying to "fix" it for me.  (i.e. for example, when I make a for-loop that iterates over an array starting at index 1 because it's deliberately avoiding the first iteration, I usually explicitly say in a comment "the fact that this doesn't start at zero isn't a mistake.  I'm skipping the first item on purpose."  Because I know that starting at one *looks* like a beginner's mistake.)

I'm hoping that using Unity5 will also make the threading smoother, allowing us access to better ways to monitor the execution "time slice" than just IPU count.

**Comment by hvacengi** [comment]
Is this an instance where compiled regex would give a noticeable performance boost?  In most benchmarks I've seen, the benefit is mostly in larger batch runs of regex matches, but I'm not sure that it requires the batch be run in sequence.  In our case, I think that the scanner basically runs through each "line" and checks it against each token until it finds a match.  That means a potentially very large number of scans.

If we wanted to improve performance further, creating an assembly with "CompileToAssembly" regexs stored in it might shift the front end load from the user's computer to our compiling computer.  It looks a little more complicated because I don't think it's in the C# build chain (for some reason the examples show creating the assembly using a console program rather than just building).  Since we don't change these regexs in operation, the change in the assembly would just have to be a step in the release and testing.

Thanks @marianoapp for pointing this out.  Even if we can improve the "feel" of the performance with threading, it would be nice to optimize some of the known sticking points too.

**Comment by Dunbaratu** [comment]
Whether the regexes under discussion here are compiled or not is the prerogative of the TinyPG code, not our code. All we do is define what the regex string looks like in the kRISC.tpg file.  TinyPG actually builds the token scanner C# code from reading that.  We could dig down into the TinyPG code to see what it's doing with it, but to change it if we don't like it would require that we fork TinyPG code and make our own version.  (Not that big a deal, as we did clone a fork of it for our own use, but that's mostly because it looked like a dead project and we didn't want to lose access to it, as we are utterly dependent on it.)

**Comment by hvacengi** [comment]
That might actually be a perfect use for us for the TinyPG fork.  If we just import the cs file that TinyPG outputs, we could easily import a dll with the compiled regexs at the same time.  I'll leave that thought to an optimization after KSP 1.1 drops though.

**Comment by Dunbaratu** [comment]
On hold until KSP 1.1, and it will likely use a completely different design than the exploratory tests that are already floating around out there about this.

**Comment by hvacengi** [comment]
Though we discussed it internally on slack, it didn't get mentioned here: KSP's selected version of Unity does not support compiled regular expressions, no matter how much I tried to trick it into working.

Fret not though, because now that we can have functions/opcodes easily tell the cpu to wait I have a plan to implement splitting these compiles pretty effectively using a threading approach.  It will steal a little bit of implementation from a swing I took at this issue a year ago.  But essentially the parsing will be able to run in parallel to everything else going on in KSP.

Not sure if we can get into the next version, but I'm gonna try to if there's time.
